### PR TITLE
RALP-4851 add screenreader focus on disabled switch states

### PR DESCRIPTION
**Pull request by LewisI224**
[RALP-4851](https://skyscanner.atlassian.net/browse/RALP-4851)
Previously when a user was reading a screenreader, it would skip over any compose BpkSwitch elements if they were in a disabled switch, this doesn't adhere to acessibility guidelines as any elements that are visible should be read out by the screenreader.

| Before | After |
|--------|------|
| <video src="https://github.com/Skyscanner/backpack-android/assets/19375170/9cb3c0c6-f2f2-40f8-a511-aeae9c8878cc"/> | <video src="https://github.com/Skyscanner/backpack-android/assets/19375170/e1b3a6ac-e83d-4842-ba43-da05ff413b37"/>

+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] Component `README.md`
+ [ ] Tests

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt [hunk 1]
```diff
@@ -73,15 +75,20 @@ fun BpkSwitch(
...
-                value = checked,
-                role = Role.Switch,
-                interactionSource = interactionSource,
-                indication = null,
-                onValueChange = onCheckedChange!!,
-            )
-        },
+        modifier = modifier
+            .applyIf(onCheckedChange != null) {
+                toggleable(
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:80]: i think actually what is missing i that we're not passing `enabled` through to the `toggleable` modifier

File: app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png
Comment (marianeum) [inline app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png:1]: any idea why these have changed? 🤔

File: .idea/deploymentTargetSelector.xml
[the remark this answers is not in the document]
Comment (LewisI224) [reply .idea/deploymentTargetSelector.xml:1]: oop. i'll revert that and add it to the git ignore

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt (hunk 1 above)
Comment (LewisI224) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:80]: how do you mean? i've tried passing it through as a parameter but it doesn't seem to have any effect, also requires the OnCheckedChange to be able to affect the disabled switches as it's not a required parameter :/

File: app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png
Comment (marianeum) [reply app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png:1]: It looks like all the text moved, which doesn't make much sense. Is your branch up to date with main?

Comment (LewisI224) [reply app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png:1]: they've actually all dissapeared now for me 👻 not sure what was going on

Comment (LewisI224) [reply app/screenshots/oss/debug/default/net.skyscanner.backpack.compose.switch.BpkSwitchTest_customContent.png:1]: ah I did run the BpkSwitch tests locally but didn't notice any snapshots get added to the changed file :/ good to know though :)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt (hunk 1 above)
Comment (LewisI224) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:80]: is this the intended behaviour you're meaning? my solution isn't the best but i can't think of.a way of using the selected semantics to improve it 🤔 <video src="https://github.com/Skyscanner/backpack-android/assets/19375170/467cc629-98f3-4e1d-bc11-7f8f4401667b"/>

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt [hunk 2]
```diff
@@ -83,10 +84,12 @@ fun BpkSwitch(
                     interactionSource = interactionSource,
                     indication = null,
                     onValueChange = onCheckedChange!!,
+                    enabled = enabled,
                 )
             }.applyIf(!enabled) {
                 semantics(mergeDescendants = true) {
                     disabled()
+                    stateDescription = if (checked) "on" else "off"
```
[the remark this answers is not in the document]
Comment (LewisI224) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:92]: `this.selected = selected` appears to always jut crash the app :/ happy to pair on this to make sure i'm not misunderstanding anything :)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt [hunk 3]
```diff
@@ -73,15 +79,25 @@ fun BpkSwitch(
...
+                    indication = null,
+                    onValueChange = onCheckedChange!!,
+                    enabled = enabled,
+                )
+            }.applyIf(onCheckedChange == null) {
+                semantics(mergeDescendants = true) {
+                    role = Role.Switch
+                    if (!enabled) {
+                        disabled()
+                        toggleableState = ToggleableState(checked)
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:97]: can you move this out of the `enabled` if statement? that way it'll also be announced for an enabled switch if there's no click listener (not that that should ever be done...)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt [hunk 4]
```diff
@@ -73,15 +79,25 @@ fun BpkSwitch(
...
+                    }
+                }.focusable()
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:99]: i don't think you actually need this

Comment (LewisI224) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:99]: without it the screenreader skips over the disabled switches :(

Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:99]: hmm odd it doesn't do for me? if you look at the implementation it actually doesn't do anything if the `enabled = false` property is set in the function, so it is not expected to be used for disabled items

Comment (LewisI224) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/switch/BpkSwitch.kt:99]: oop it's my laziness that made me think it wasn't working. i've been using arrow keys to move the screenreader focus as using the touchpad to swipe properly has been really temperamental. 

When swiping properly the screenreader doesn't skip 👍

**Review by marianeum — APPROVED** [review]

**Pull request merged**